blockchain/state_manager/src/lib.rs

```diff
@@ -117,6 +131,98 @@ where
+            let cid_pair = (
+                tipset.parent_state().clone(),
+                message_receipts.message_receipts().clone(),
+            );
+            self.cache.insert(tipset.key().clone(), cid_pair.clone());
+            return Ok(cid_pair);
+        }
+
+        let block_headers = tipset.blocks().to_vec();
+        // generic constants are not implemented yet this is a lowcost method for now
+        let cid_pair = self.compute_tipset_state(&block_headers, &HashMap::new())?;
+        self.cache.insert(tipset.key().clone(), cid_pair.clone());
+        Ok(cid_pair)
+    }
+
+    pub fn compute_tipset_state<'a>(
+        &'a self,
+        blocks: &[BlockHeader],
+        forks_at_heights: &'a HashMap<ChainEpoch, ForkFunctions<'a, DB>>,
+    ) -> Result<(Cid, Cid), Box<dyn StdError>> {
+        trace!("compute tipset state");
+        if blocks.len() > 2
+            && blocks
+                .iter()
+                .zip(blocks.iter().skip(0))
+                .any(|(a, b)| a.miner_address() == b.miner_address())
+        {
+            // Duplicate Minor found
+            return Err(Box::new(Error::Other(
+                "Could not get message receipts".to_string(),
+            )));
+        }
+
+        let parents_cid = &blocks
+            .first()
+            .ok_or_else(|| Error::Other("Could not get message receipts".to_string()))?
+            .parents()
+            .cids;
+
+        if !parents_cid.is_empty() {
+            let parents_first = parents_cid
+                .first()
+                .ok_or_else(|| Error::Other("Could not get message receipts".to_string()))?;
+            self.bs
+                .get(parents_first)?
+                .ok_or_else(|| Error::Other("Could not get message receipts".to_string()))?;
+            // handle state forks
+            let func_to_execute = forks_at_heights
+                .get(
+                    &blocks
+                        .first()
+                        .ok_or_else(|| Error::Other("Could not get message receipts".to_string()))?
+                        .epoch(),
+                )
+                .ok_or_else(|| Error::Other("Could not get message receipts".to_string()))?;
+            func_to_execute(&self, parents_first.clone())?;
+        };
+
+        let chain_store = ChainStore::new(self.bs.clone());
+        let blocks = blocks
+            .iter()
+            .map::<Result<Block, Box<dyn StdError>>, _>(|s: &BlockHeader| {
+                let (bls_messages, secp_messages) = chain_store.messages(&s)?;
+                Ok(Block {
+                    header: s.clone(),
+                    bls_messages,
+                    secp_messages,
+                })
+            })
+            .collect::<Result<Vec<Block>, _>>()?;
+        let full_tipset = FullTipset::new(blocks)?;
+        // convert tipset to fulltipset
+        self.apply_blocks(&full_tipset)
+    }
```

Comment:
Where is this logic coming from? I can't seem to map it to anything

Reply:
computeTipsetState in lotus